packages/plugin-ext/src/main/browser/languages-main.ts

```diff
@@ -698,35 +693,56 @@ export class LanguagesMainImpl implements LanguagesMain {
     }
 }
 
-function reviveMarker(marker: MarkerData): monaco.editor.IMarkerData {
-    const monacoMarker: monaco.editor.IMarkerData = {
+function reviveMarker(marker: MarkerData): vst.Diagnostic {
+    const monacoMarker: vst.Diagnostic = {
         code: marker.code,
-        severity: marker.severity,
+        severity: reviveSeverity(marker.severity),
+        range: reviveRange(marker.startLineNumber, marker.startColumn, marker.endLineNumber, marker.endColumn),
         message: marker.message,
         source: marker.source,
-        startLineNumber: marker.startLineNumber,
-        startColumn: marker.startColumn,
-        endLineNumber: marker.endLineNumber,
-        endColumn: marker.endColumn,
         relatedInformation: undefined
     };
+
     if (marker.relatedInformation) {
-        monacoMarker.relatedInformation = [];
-        for (const ri of marker.relatedInformation) {
-            monacoMarker.relatedInformation.push({
-                resource: monaco.Uri.revive(ri.resource),
-                message: ri.message,
-                startLineNumber: ri.startLineNumber,
-                startColumn: ri.startColumn,
-                endLineNumber: ri.endLineNumber,
-                endColumn: ri.endColumn
-            });
-        }
+        monacoMarker.relatedInformation = marker.relatedInformation.map(reviveRelated);
     }
 
     return monacoMarker;
 }
 
+function reviveSeverity(severity: MarkerSeverity): vst.DiagnosticSeverity {
+    switch (severity) {
+        case MarkerSeverity.Error: return vst.DiagnosticSeverity.Error;
+        case MarkerSeverity.Warning: return vst.DiagnosticSeverity.Warning;
+        case MarkerSeverity.Info: return vst.DiagnosticSeverity.Information;
+        case MarkerSeverity.Hint: return vst.DiagnosticSeverity.Hint;
+    }
+}
+
+function reviveRange(startLine: number, startColumn: number, endLine: number, endColumn: number): vst.Range {
+    // note: language server range is 0-based, marker is 1-based, so need to deduct 1 here
+    return {
+        start: {
+            line: startLine - 1,
```

Comment:
If possible please use https://github.com/theia-ide/theia/blob/master/packages/plugin-ext/src/plugin/type-converters.ts for such conversions.

Reply:
again, yes
type-converter is converting both ways (and the name is generic enough and is in an upper package to be widely used)
example for Position: https://github.com/theia-ide/theia/blob/master/packages/plugin-ext/src/plugin/type-converters.ts#L96-L102
it has `from` and `to`

Reply:
Sorry, but no: the file you refer to is in the "plugin" subtree and converts between plugin API objects and the main/ext API. The conversion here happens on the other side of the main/ext API and converts to objects from the module 'vscode-languageserver-types'.
Also, the this file already contains a number of functions that convert from main/ext types to monaco or theia internal types. So my change here conforms to the already existing code.

Reply:
type-converter is in plug-in subtree and you're converting stuff in the file `packages/plugin-ext/src/main/browser/languages-main.ts` so it's the same subtree.

It's converting types coming from `vscode-languageserver-types'

you introduced a local scope
``` import * as vst from 'vscode-languageserver-types';``` to convert data to the same origin, 
so it should be in `https://github.com/theia-ide/theia/blob/master/packages/plugin-ext/src/plugin/type-converters.ts`

Reply:
No, it's in the "main" subtree. I understand what you're saying, but I disagree with your argument.

Reply:
@tsmaeder `src/main` subtree can import anything from `src/plugin`

Reply:
Yes, technically, it can, but if the type converter provides services to both the plugin and main side of things, it should be moved to the "common" folder, otherwise, why do we even have the subfolders? I feel that is outside of the scope of this PR.
Anyway, we're having a stylistic difference here we're not going to resolve amicably. How should we proceed?

Reply:
@akosyakov you mean this line https://github.com/theia-ide/theia/blob/master/packages/plugin-ext/src/plugin/type-converters.ts#L27 ? (imports of vscode-types ) ?

or that https://github.com/theia-ide/theia/blob/a83fb71067315159706ab274f31e2fa5c4393644/packages/plugin-ext/src/main/browser/languages-main.ts (language main) can't access `type-converter` ?

I'm also fine with your proposal

Reply:
> or that https://github.com/theia-ide/theia/blob/a83fb71067315159706ab274f31e2fa5c4393644/packages/plugin-ext/src/main/browser/languages-main.ts (language main) can't access `type-converter` ?

I meant that when we import `type-converters` in main, it [pulls JS files like `types-impl`](https://github.com/theia-ide/theia/blob/403cc2f60db05270bf2512cc992f77f31a46ecd9/packages/plugin-ext/src/plugin/type-converters.ts#L20) and other internal code of the host plugin process which gets bundled by webpack or loaded by Node process. Is it really necessary, expected?

Reply:
I've opened an issue for it - https://github.com/theia-ide/theia/issues/4271

Thinking more about it, I am not sure that we can do all conversions in one file. There are internal VS Code types for plugin module and internal Monaco types for main module, trying to make conversion between them in one file will always cause such issues. Common converter can only depend on `theia` namespace, LSP and plugin DTO types.